### DISCUSSION TITLE
How to get client IP in Quarkus JAX-RS service @ApplicationScoped?

### DISCUSSION THREAD

**Discussion by dhoffer** [quarkusio/quarkus#27451] [Community]
I have tried injecting HttpServletRequest but get CDI build errors.  I have tried @Context HttpServletRequest in the service method and I get rutime error of RESTEASY003880: Unable to find contextual data of type: javax.servlet.http.HttpServletRequest

How can I get the client IP calling the services?

-Dave

**Comment by mkouba** [comment]
> I have tried injecting HttpServletRequest but get CDI build errors

What errors did you get? It should be possible to `@Inject io.vertx.core.http.HttpServerRequest`.

Hm, I can see you're trying to use the servlet API. Do you use the undertow extension?

**Comment by jimbogithub** [comment]
I believe the following is the preferred way if using the reactive libraries:
```
@Context
io.vertx.core.http.HttpServerRequest httpRequest;
```
Works for us in `@ApplicationScope`, `@RequestScope` and `@javax.ws.rs.ext.Provider` (e.g. filters).

**Comment by dhoffer** [comment]
I was using javax.servlet.http.HttpServletRequest.  I just tried @Inject io.vertx.core.http.HttpServerRequest and it seems to work, I'll investigate this more.  We are not using reactive JAX-RS libraries, so is @Inject preferred over @Context?

**Comment by pedropds** [comment]
Just had this problem myself. Injecting `HttpServerRequest` would work on the Controller layer, but on on the Service layer, it was a null object. Which is kinda lame, since it appears to lose the context, and i would have to pass this object to every service method that needed this.
Just inject `RoutingContext` instead. From the routing context you can get the `HttpServerRequest` by using the `RoutingContext.request() ` function. Good thing about this is that you can inject this into the Service layer as well and will still have all the context, given that the Service was called by a Controller. 

Hope this helps!

**Reply by mkouba** [reply]
>  Good thing about this is that you can inject this into the Service layer as well and will still have all the context, given that the Service was called by a Controller.

So this is weird because both the `io.vertx.ext.web.RoutingContext` and `io.vertx.core.http.HttpServerRequest` beans are `@RequestScoped`. In fact, the producer for the latter injects the first one - see also https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/CurrentRequestProducer.java#L18-L23. In other words, `@Inject` should work the same way for both, whereas `@Context` only works in JAX-RS resources.
